Chapter_4_Advanced_Structuring_and_Execution.md

```diff
@@ -0,0 +1,599 @@
+# 4 Advanced Structuring and Execution
+
+As a Robot Framework automation project expands, the increasing number of tests|tasks adds complexity to the project.
+This chapter explores advanced structuring and execution techniques to effectively manage this complexity and control the execution flow.
+
+We will cover methods for error handling and cleaning up after failed tests|tasks using **Teardowns**, as well as preparing individual or multiple suites and tests|tasks for execution with **Setups**.
+Additionally, filtering subsets of tests|tasks based on tags will be discussed, which is essential for managing test|task execution efficiently.
+
+
+
+
+## 4.1 Setups (Suite, Test|Task, Keyword)
+
+> [!IMPORTANT]
+> LO-XX Recall the purpose and benefits of Setups in Robot Framework (K1)
+>
+> LO-XX Understand the different levels where and how a Setup can be defined (K2)
+
+
+Setups in Robot Framework are used to prepare the environment or system for execution or to verify that the requirements/preconditions needed for execution are met.
+They can be defined at the suite, test|task, or keyword level and are executed before the respective scope begins execution.
+
+A **Setup** is a single keyword with potential argument values that is called before all other keywords; or before tests|tasks in Suite Setup.
+
+Examples of typical use cases for Setups are:
+- Establishing connections to databases or services.
+- Initializing test data or configurations.
+- Setting the system under test to a known state.
+- Logging into applications or systems.
+- Navigating to the feature under test.
+
+
+
+### 4.1.1 Suite Setup
+
+> [!IMPORTANT]
+> LO-XX Recall key characteristics, benefits, and syntax of Suite Setup (K1)
+>
+> LO-XX Understand when Suite Setup is executed and used (K2)
+
+A **Suite Setup** is executed before any tests|tasks or child suites within the suite are run.
+It is used to prepare the environment or perform actions that need to occur before the entire suite runs.
+Since it is only executed once before all tests|tasks or child suites, it can save time, rather than executing the action for each test|task individually.
+
+**Key characteristics of Suite Setup:**
+- Executed before any tests|tasks and child suites in the suite.
+- If the Suite Setup fails, all tests|tasks in the suite and its child suites are marked as failed, and they are not executed.
+- Ideal for checking **preconditions** that must be met before running the tests|tasks.
+- Logged in the execution log as a separate section, indicating the setup status.
+
+**Typical use cases:**
+- Ensuring that the environment is ready for execution.
+- Starting services or applications required for the suite.
+- Preparing a system under test to meet the suite's requirements.
```

Comment:
I would leave out "under test" as this focusses on tests, but it's the same for tasks

```suggestion
- Preparing a system to meet the suite's requirements.
```